### PR TITLE
Develop/cg token

### PR DESCRIPTION
**Pull request by devcrazygit**

### CODE REVIEW

File: contracts/CryptoGoods.sol [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+        require(_whiteList[msg.sender] > 0, "You are not allowed");
+        require(
+            _whiteList[msg.sender] >= numberOfTokens,
+            "Exceed max whitelist available"
+        );
+
+        uint256 ts = totalSupply();
+        require(
+            ts + numberOfTokens <= MAX_PRESALE_SUPPLY,
+            "Purchase would exeed max tokens"
```
Comment (glinda93) [inline contracts/CryptoGoods.sol:118]: typo

File: contracts/CryptoGoods.sol [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+        );
+        require(
+            currentPrice * numberOfTokens <= msg.value,
```
Comment (glinda93) [inline contracts/CryptoGoods.sol:121]: Use `_msgValue()` instead of `msg.value`

File: contracts/CryptoGoods.sol [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+        for (uint8 i = 0; i < numberOfTokens; i++) {
+            uint256 tokenId = _mintTo(msg.sender);
+            _setTokenURI(tokenId, string(abi.encodePacked(BASE_URI, tokenId)));
+        }
+        _whiteList[msg.sender] -= numberOfTokens;
+    }
+
+    function withdraw() public onlyOwner returns (bool) {
+        uint256 balance = address(this).balance;
+        if (payable(msg.sender).send(balance)) {
```
Comment (glinda93) [inline contracts/CryptoGoods.sol:133]: I don't think it's a good way to send eth.

https://solidity-by-example.org/sending-ether/

File: contracts/CryptoGoods.sol [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+            "Ether is not enough"
+        );
+        for (uint8 i = 0; i < numberOfTokens; i++) {
+            uint256 tokenId = _mintTo(msg.sender);
+            _setTokenURI(tokenId, string(abi.encodePacked(BASE_URI, tokenId)));
+        }
+        _whiteList[msg.sender] -= numberOfTokens;
+    }
+
+    function withdraw() public onlyOwner returns (bool) {
```
Comment (glinda93) [inline contracts/CryptoGoods.sol:131]: Why do we need `withdraw` function in NFT?

File: contracts/CryptoGoods.sol [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+        onlyOwner
+    {
+        _setCurrentMarketWithPrice(marketStatus, 0);
+    }
+
+    /**
+     * @dev set current nft market status (whitelist, presale, sale)
+     * owner can manually change status
+     */
+    function setCurrentMarketWithPrice(MarketStatus marketStatus, uint256 price)
```
Comment (glinda93) [inline contracts/CryptoGoods.sol:61]: Does it have to be renamed to `setCurrentMarketStatusWithPrice` ?

File: contracts/CryptoGoods.sol [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+        external
+        onlyOwner
+    {
+        _setCurrentMarketWithPrice(marketStatus, price);
```
Comment (glinda93) [inline contracts/CryptoGoods.sol:65]: Same here

File: contracts/CryptoGoods.sol [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+
+    /**
+     * @dev mint at sale, giveaway status
+     */
+    function mint() public payable {
+        require(
+            currentMarketStatus != MarketStatus.NONE,
+            "Market is not opened yet"
+        );
+        if (currentMarketStatus == MarketStatus.PRESALE) {
```
Comment (glinda93) [inline contracts/CryptoGoods.sol:93]: I don't understand what's happening here.

Can you elaborate?

File: test/index.ts [hunk 8]
```diff
@@ -1,19 +1,261 @@
...
+        })
+      ).to.be.revertedWith("Ether is not enough");
+    });
+    it("should allow users to mint at sale price", async function () {
+      const initialCgToken = await CryptoGoodsToken.balanceOf(addr1.address);
+      await CryptoGoodsToken.connect(addr1).mint({ value: salePrice });
+      const finalCgToken = await CryptoGoodsToken.balanceOf(addr1.address);
+      expect(finalCgToken.sub(initialCgToken)).to.be.equal(1);
+    });
+    // it("should be changed into giveaway status", async function() {
```
Comment (glinda93) [inline test/index.ts:257]: Remove zombie codes here

File: test/index.ts [hunk 9]
```diff
@@ -1,19 +1,261 @@
+import { SignerWithAddress } from "@nomiclabs/hardhat-ethers/signers";
```
Comment (glinda93) [inline test/index.ts:1]: Please use this pattern when writing tests:

```
describe('Class', () => {
   describe('method', () => {
     it('should do something', () => {
        expect(true).toBe(true);
     });
     // optional
     describe('when some condition matches', () => {
       it('should do another thing', () => {
          
       });
     });
   });
});
```

**Review by glinda93 — CHANGES_REQUESTED** [review]

File: contracts/CryptoGoods.sol (hunk 7 above)
Comment (devcrazygit) [reply contracts/CryptoGoods.sol:93]: Whitelist users can use both of `mintAtPresale` and `mint`.
One for mint in bulk, and another for mint one by one. It depends on the client implementation.

**Pull request merged**